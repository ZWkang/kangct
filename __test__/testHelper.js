// import React from 'react';
import { mount, shallow, render } from 'enzyme';

function setUp(type, ...arg) {
  const [Component, props = {}, children = ''] = arg;
  if (~['mount', 'shallow', 'render'].indexOf(type)) {
    return global[type](<Component {...props}>{children}</Component>);
  }
}

export function AutoMath(Component, props = {}, children = '') {
  const wrapper = setUp('render', Component, props, children);
  expect(wrapper).toMatchSnapshot();
}

export function ComponentRenderWithoutError(Component, props = {}, children = '') {
  const wrapper = setUp('mount', Component, props, children);
  expect(() => {
    wrapper.setProps({});
    wrapper.unmount();
  }).not.toThrow();
}

export function childrenShouldBeProps(Component, children) {
  const wrapper = setUp('mount', Component, {}, children);

  expect(wrapper.get(0).props.children).toEqual(children);
}

export function setup(Component, props) {
  const wrapper = setUp(Component, props, null);
  return {
    wrapper,
    props
  };
}
