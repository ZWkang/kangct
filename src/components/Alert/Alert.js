import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import is from 'styled-is';

const colorMaps = {
  success: {
    bgColor: 'rgb(212, 237, 217);',
    color: 'rgb(20, 86, 35);',
    bdColor: 'rgb(194, 230, 202);'
  },
  info: {
    bgColor: 'rgb(208, 236, 240);',
    color: 'rgb(11, 84, 95);',
    bdColor: 'rgb(190, 228, 235);'
  },
  danger: {
    bgColor: 'rgb(248, 214, 217);',
    color: 'rgb(114, 27, 35);',
    bdColor: 'rgb(245, 198, 202);'
  },
  warning: {
    bgColor: 'rgb(255, 242, 205);',
    color: 'rgb(132, 100, 3);',
    bdColor: 'rgb(255, 237, 185);'
  },
  secondary: {
    bgColor: 'rgb(225, 227, 229);',
    color: 'rgb(56, 60, 65);',
    bdColor: 'rgb(225, 227, 229);'
  }
};

const handleAlertTypeColorChoose = (maps) => (name) => (props) => {
  let mapKey = props[name];
  const mapKeys = Object.keys(maps);
  if (~mapKeys.indexOf(mapKey) >= 0) {
    mapKey = mapKeys[0];
  }
  return css`
    ${Object.keys(maps[mapKey])
      .map((item, index, source) => {
        if (item === 'bgColor') {
          return `background-color: ${maps[mapKey][item]}`;
        }
        if (item === 'color') {
          return `color: ${maps[mapKey][item]}`;
        }
        if (item === 'bdColor') {
          return `border-color: ${maps[mapKey][item]}`;
        }
      })
      .filter(Boolean)
      .join('')}
  `;
};

const AlertWrapper = styled.div`
  position: relative;
  margin-bottom: 1rem;
  color: rgb(98, 37, 26);
  background-color: rgb(241, 218, 214);
  padding: 0.75rem 1.25rem;
  border-width: 1px;
  border-style: solid;
  border-image: initial;
  border-radius: 0.25rem;
  border-color: rgb(236, 204, 197);
  ${is('type')`
    ${handleAlertTypeColorChoose(colorMaps)('type')}
  `}
`;
const AlertClose = styled.i`
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  font-size: 14px;
  cursor: pointer;
  right: 16px;
`;
const AlertContentWrapper = styled.div`
  display: flex;
  flex-direction: column;
`;
const AlertTitle = styled.div`
  font-size: 1rem;
  margin-bottom: 4px;
`;
const AlertMessage = styled.div`
  font-size: 12px;
  margin-bottom: 4px;
`;

/**
 * autoClose
 * closeTime
 * type
 * title
 * message
 * canClose
 * show
 * onclose
 */
class Alert extends Component {
  constructor(props) {
    super(props);
    this.state = {
      show: props.show
    };
  }
  static defaultProps = {
    canClose: false,
    autoClose: false,
    type: 'info',
    title: '',
    show: true,
    closeTime: 2000
  };
  componentWillReceiveProps(nextProps) {
    if (nextProps.show !== this.state.show) {
      this.setState({
        show: nextProps.show
      });
    }
  }
  componentDidMount() {
    const { autoClose, closeTime = 2000 } = this.props;

    if (autoClose === true) {
      setTimeout(() => {
        this.handleAlertClose();
      }, closeTime);
    }
  }
  handleAlertClose = (e) => {
    const { onClose } = this.props;
    this.setState(
      {
        show: false
      },
      () => {
        onClose && onClose(e, this.state, this.props, this);
      }
    );
  };

  render() {
    const { title, message, canClose, children, type } = this.props;
    const { show } = this.state;
    if (!show) return null;
    return (
      <AlertWrapper type={type}>
        <AlertContentWrapper>
          <AlertTitle>{title}</AlertTitle>
          <AlertMessage>{message ? message : children}</AlertMessage>
        </AlertContentWrapper>
        {canClose ? <AlertClose onClick={this.handleAlertClose}>X</AlertClose> : null}
      </AlertWrapper>
    );
  }
}

Alert.propTypes = {
  canClose: PropTypes.bool,
  autoClose: PropTypes.bool,
  type: PropTypes.string,
  title: PropTypes.string,
  show: PropTypes.bool,
  closeTime: PropTypes.number
};
export default Alert;
