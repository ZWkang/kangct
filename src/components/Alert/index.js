import Alert from './Alert'

export { Alert as default }