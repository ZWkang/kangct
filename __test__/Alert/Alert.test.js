// import React from 'react';
// import { mount, shallow, render } from 'enzyme';
import Alert from '@app/Alert/index.js';
import { AutoMath, ComponentRenderWithoutError } from '../testHelper';
import 'jest-styled-components';

// https://github.com/airbnb/enzyme/issues/1798 dive vs shallow
describe('Alert', () => {
  describe('behavior test', () => {
    it('render without error', () => {
      ComponentRenderWithoutError(Alert);
    });
    it('test props snapshot', () => {
      AutoMath(Alert, { type: 'success' }, 'here is a test alert');
    });
    it('test props type should be work', () => {
      const wrapper = mount(<Alert type="success">here is a test alert</Alert>);
      expect(wrapper.find('Alert')).toHaveStyleRule('color', 'rgb(20,86,35)');
    });

    it('test click close test', () => {
      const onClose = jest.fn();
      const wrapper = mount(
        <Alert canClose type="success" onClose={onClose}>
          here is a test alert
        </Alert>
      );
      wrapper.find('i').simulate('click');
      expect(onClose).toBeCalled();
      expect(wrapper.html()).toBe(null);
    });

    it('title should be render', () => {
      const wrapper = shallow(<Alert title="title test" />);
      expect(wrapper.text()).toEqual('title test');
    });

    it('test auto close', () => {
      jest.useFakeTimers();
      const onClose = jest.fn();
      const wrapper = shallow(
        <Alert autoClose closeTime={2000} title="title test" onClose={onClose} />
      );
      jest.advanceTimersByTime(3000);
      expect(setTimeout).toHaveBeenCalledTimes(1);
      expect(onClose).toBeCalled();
    });

    it('show can control render', () => {
      const wrapper = shallow(<Alert show={false} title="title test" />);
      expect(wrapper.type()).toBeNull();
      expect(wrapper.get(0)).toBeFalsy();
      // test null
      // https://stackoverflow.com/questions/47259061/jest-enzyme-test-a-react-component-that-returns-null-in-render-method
      wrapper.setProps({ show: true });
      expect(wrapper.get(0)).not.toBeFalsy();
      expect(wrapper.type()).not.toBeNull();
    });
  });
});
