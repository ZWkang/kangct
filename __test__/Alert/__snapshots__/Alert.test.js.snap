// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert behavior test test props snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "sc-bdVaJa bIfvBd",
      "type": "success",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "sc-htpNat dePyzG",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "sc-bxivhb ebGWgl",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "sc-ifAKCX iJCkad",
              },
              "children": Array [
                Object {
                  "data": "here is a test alert",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "sc-ifAKCX iJCkad",
            },
            "children": Array [
              Object {
                "data": "here is a test alert",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "sc-bxivhb ebGWgl",
              },
              "children": Array [],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "type": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "type": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
